File: 0173-binary-search-tree-iterator/0173-binary-search-tree-iterator.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class BSTIterator {
private:
    void findinorder(TreeNode* root,vector<TreeNode*> &inorder){
        if(!root) return;
        findinorder(root->left,inorder);
        inorder.push_back(root);
        findinorder(root->right,inorder);
    }

public:

    vector<TreeNode*> inorder;
    int index;
    BSTIterator(TreeNode* root) {
        index=-1;
        findinorder(root,inorder);
    }
    
    int next() {
        return inorder[++index]->val;
    }
    
    bool hasNext() {
        return index+1<inorder.size();
    }
};

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator* obj = new BSTIterator(root);
 * int param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */

File: 0173-binary-search-tree-iterator/0173-binary-search-tree-iterator.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

from collections import deque

class BSTIterator:
    def inorder(self,root,order):
        if not root:
            return

        self.inorder(root.left,order)
        order.append(root.val)
        self.inorder(root.right,order)

        return order

    def __init__(self, root: Optional[TreeNode]):
        self.order = self.inorder(root,[])
        self.q = deque(self.order)

    def next(self) -> int:
        node = self.q.popleft()
        return node

    def hasNext(self) -> bool:
        return len(self.q) != 0



# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()